File: 규진/프로그래머스/week16/징검다리_건너기.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <iostream>

using namespace std;
bool binarySearch(int n, int k, vector<int> stones) {
    int cnt = 0;
    for (int i = 0; i < stones.size(); i++) {
        if (n >= stones[i])
            cnt++;
        else
            cnt = 0;
        if (cnt >= k) {
            return false;
        }
    }
    return true;
}

int solution(vector<int> stones, int k) {
    int answer = 0;
    int left = 1;
    int right = *max_element(stones.begin(), stones.end());
    
    while (left <= right) {
        int mid = (left + right) / 2;
        if (binarySearch(mid, k, stones)) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }
    answer = left;
    return answer;
}

File: 명진/programmers/징검다리 건너기.py
def solution(stones, k):
    def check(x):
        jump = 0
        for st in stones:
            if st - x + 1 <= 0:
                jump += 1
                if jump >= k:
                    return False
            else:
                jump = 0
        return True

    l, r = 1, 200_000_000
    while l + 1 < r:
        m = (l + r) // 2
        if check(m):
            l = m
        else:
            r = m

    return l
